contracts/oracle/src/contract/oracle/mod.rs

```diff
@@ -155,3 +155,214 @@ where
+        };
+
+        let mut storage: MockStorage = MockStorage::new();
+
+        let price_config: PriceConfig = PriceConfig::new(
+            Percent::HUNDRED,
+            Duration::from_secs(SAMPLE_PERIOD_SECS),
+            1,
+            Percent::HUNDRED,
+        );
+
+        let mut now_seconds: u64 = 1;
+
+        init(&mut storage, &price_config);
+
+        add_higher_alarm(&mut storage);
+
+        feed_below_price(&price_config, &mut storage, now_seconds);
+
+        now_seconds += 1;
+
+        dispatch_1(&mut storage, now_seconds);
+
+        add_lower_alarm(&mut storage);
+
+        now_seconds += 1;
+
+        for _ in 0..SAMPLE_PERIOD_SECS.checked_add(1).unwrap() {
+            dispatch_0(&mut storage, now_seconds);
+
+            now_seconds += 1;
+        }
+
+        feed_normal_price(&price_config, &mut storage, now_seconds);
+
+        now_seconds += 1;
+
+        dispatch_1(&mut storage, now_seconds);
+
+        now_seconds += 1;
+
+        // Bug happens on this step.
+        dispatch_0(&mut storage, now_seconds);
+
+        feed_below_price(&price_config, &mut storage, now_seconds);
+
+        now_seconds += 1;
+
+        dispatch_0(&mut storage, now_seconds);
+    }
+
+    #[track_caller]
+    fn init(storage: &mut dyn Storage, price_config: &PriceConfig) {
+        Feeders::try_register(
+            DepsMut {
+                storage,
+                api: &MockApi::default(),
+                querier: QuerierWrapper::new(&MockQuerier::<Empty>::new(&[])),
+            },
+            String::from("feeder"),
+        )
+        .unwrap();
+
+        Config::new(String::from(BaseCurrency::TICKER), price_config.clone())
+            .store(storage)
+            .unwrap();
+
+        SupportedPairs::<BaseCurrency>::new(
+            swap_tree!({ base: Usdc::TICKER }, (1, Nls::TICKER)).into_tree(),
+        )
+        .unwrap()
+        .save(storage)
+        .unwrap();
+    }
+
+    #[track_caller]
+    fn add_alarm(storage: &mut dyn Storage, below_nls_to_2: Amount, above_1_to_usdc: Amount) {
+        let storage: &mut dyn Storage = storage;
+
+        let mut alarms: MarketAlarms<'_, &mut dyn Storage> = MarketAlarms::new(storage);
+
+        alarms
+            .try_add_price_alarm::<BaseCurrency>(
+                Addr::unchecked("1"),
+                Alarm::new(
+                    SpotPrice::new(NlsCoin::new(below_nls_to_2).into(), UsdcCoin::new(2).into()),
+                    Some(SpotPrice::new(
+                        NlsCoin::new(1).into(),
+                        UsdcCoin::new(above_1_to_usdc).into(),
+                    )),
+                ),
+            )
+            .unwrap();
+    }
+
+    #[track_caller]
+    fn add_higher_alarm(storage: &mut dyn Storage) {
+        add_alarm(storage, 1, 3)
+    }
+
+    #[track_caller]
+    fn add_lower_alarm(storage: &mut dyn Storage) {
+        add_alarm(storage, 2, 2)
+    }
+
```

Comment:
replace this code with price constants and pass them to `add_alarm` not `Amount`-s